addons/addon-user-id-upgrade/packages/user-id-upgrade-post-deployment/lib/steps/upgrade-to-user-id.js

```diff
@@ -0,0 +1,179 @@
+
+      // Do the rest of the tables
+      const compositeError = new CompositeError();
+      // This function is a worker function, it uses the table migrator object
+      // to do the migration of a table.
+      const doWork = async item => {
+        try {
+          const migrator = new TableMigrator({
+            upgradeUserIdService: this,
+            uidLookup,
+            // table is the canonical table name. Using the canonical name, we can deduce
+            // the name of the source and target tables.
+            table: item.name,
+            keyNames: item.keys, // the key names for the target table
+            transforms: item.transforms, // a list of functions that transforms a row
+            batchSize: item.batchSize || 10, // default is 10 rows at a time
+          });
+          await migrator.run();
+        } catch (error) {
+          compositeError.addError(error);
+        }
+      };
+
+      const tables = tablesFn();
+      this.log.info(prefix(`About to migrate the remaining ${_.size(tables)} tables`));
+
+      // Now, lets do them in batches (5 tables at a time)
+      const batches = _.chunk(tables, 5);
+      while (!_.isEmpty(batches)) {
+        const batch = batches.shift();
+        await Promise.all(_.map(batch, doWork));
+        if (compositeError.hasErrors) throw compositeError;
+      }
+
+      await this.saveDeploymentItem({ id: 'status', value: 'completed' });
+      this.log.info(prefix('Completed'));
+    } catch (error) {
+      this.log.info(prefix('Failed'));
+      logError(error, this.log);
+      throw new Error('Upgrading to user id step has errors. See the previous log message for more details.');
+    } finally {
+      this.metrics.end();
+      // We don't want to save the metrics in the database if the upgrade is skipped
+      if (shouldUpgrade) {
+        await this.metrics.save();
+        await this.report.save();
+      }
+      const time = this.metrics.toMilliseconds();
+      this.log.info({ msg: prefix(`took ${time} ms`), runTime: time });
+    }
+  }
+
+  // Determines if the upgrade should be attempted. The logic is as follows:
+  // - Check if the old users table has an entry for 'root', if it does NOT, then
+  //   it means that this is a new installation and there is no need to do the upgrade.
+  // - Check if the deployment store has 'completed' for user-id-upgrade-v1
+  //   if so, then there is no need to do the upgrade.
+  // - If none of the above, then we need to do the upgrade.
+  async shouldUpgrade() {
+    // Do we have a root user in the depreciated DbUsers table
+    const depreciatedUsersTable = this.getDepreciatedTableName(settingKeys.users);
+    const tableReporter = this.report.getTableReport(depreciatedUsersTable);
+    const getItem = new GetItem(depreciatedUsersTable, this.api);
+    const key = {
+      username: toStringAttributeValue('root'),
+      ns: toStringAttributeValue('internal'),
+    };
```

Comment:
Just to confirm this is a guarantee that this user will be in an unmigrated table? I feel a safer approach is to scan and if there are entries then migrate

Reply:
Yes, the root user is created as part of the installation, and it is guaranteed to be there for installations that need to be migrated.  However, I agree with you that a scan might be a safer approach. The trick is that this might require me to do more testing for the new change and I am not sure we have any more time dedicated to this work.

Reply:
Fair -- I this approach is fine